src/wagtail_ai/prompts.py

```diff
@@ -1,82 +1,29 @@
-        pass
-
-    id: int
-    label: str
-    prompt: str
-    backend: str = "default"
-    description: str = ""
-    method: Union[str, Method] = Method.REPLACE
-
-    def __post_init__(self):
-        self.prompt = inspect.cleandoc(self.prompt)
-        self.method = Prompt.Method(self.method)
-
-    def as_dict(self) -> dict:
-        return {
-            "id": self.id,
-            "label": self.label,
-            "description": self.description,
-            "prompt": self.prompt,
-            "method": Prompt.Method(self.method).value,
-        }
-
+from .models import Prompt
 
 DEFAULT_PROMPTS = [
     {
         "label": "AI Correction",
         "description": "Correct grammar and spelling",
-        "prompt": """You are assisting a user in writing content for their website.
-            The user has provided some text (following the colon).
-            Return the provided text but with corrected grammar, spelling and
-            punctuation.
-            Do not add additional punctuation, quotation marks or change any words:""",
+        "prompt": (
+            "You are assisting a user in writing content for their website. "
+            "The user has provided some text (following the colon). "
+            "Return the provided text but with corrected grammar, spelling and punctuation. "
+            "Do not add additional punctuation, quotation marks or change any words:"
+        ),
         "method": "replace",
     },
     {
         "label": "AI Completion",
         "description": "Get help writing more content based on what you've written",
-        "prompt": """You are assisting a user in writing content for their website.
-            The user has provided some initial text (following the colon).
-            Assist the user in writing the remaining content:""",
+        "prompt": (
+            "You are assisting a user in writing content for their website. "
+            "The user has provided some initial text (following the colon). "
+            "Assist the user in writing the remaining content:"
+        ),
         "method": "append",
     },
 ]
 
 
-def get_prompts():
-    try:
-        return [
-            Prompt(**prompt, id=idx)
-            for idx, prompt in enumerate(
-                getattr(settings, "WAGTAIL_AI_PROMPTS", []) or DEFAULT_PROMPTS
-            )
-        ]
-    except TypeError as e:
-        raise ImproperlyConfigured(
-            """WAGTAIL_AI_PROMPTS must be a list of dictionaries, where each dictionary
-            has a 'label', 'description', 'prompt' and 'method' key.
-            The 'method' must be one of 'append' or 'replace'."""
-        ) from e
-
-
-def get_prompt_by_id(id: int) -> Prompt:
-    for prompt in get_prompts():
-        if prompt.id == id:
-            return prompt
-    raise Prompt.DoesNotExist(f"Can't get prompt by id: {id}")
+def get_prompts() -> list[Prompt]:
+    return [prompt.as_dict() for prompt in Prompt.objects.all()]  # type: ignore
```

Comment:
**Issue:** Can we selectively use what the field needs, not dump everything? May be revealing sensitive information and more fundamentally the numeric PKs that may be used to learn about the database structure.

Reply:
Wouldn't a a DRF serializer make more sense here?

Something like:
```
class PromptSerializer(serializers.Serializer):
    uuid = serializers.UUIDField()
    label = serializers.CharField()
    description = serializers.CharField()
    prompt = serializers.CharField()
    method = serializers.CharField()

def get_prompts_for_api():
    prompts = Prompt.objects.all()
    serializer = PromptSerializer(prompts, many=True)
    return serializer.data
```

Reply:
Apologies. I did not realise your as_dict method does that already (selectively picking fields). In some ways, the code is a bit obscure because the calling code reads to me as "dump a dict with all the fields" I guess.

```python
  def as_dict(self) -> dict:
        return {
            "uuid": str(self.uuid),
            "label": self.label,
            "description": self.description,
            "prompt": self.prompt_value,
            "method": self.Method(self.method).value,
        }
```

If the method is not used anywhere else, I think it'd make more sense to make the code more explicit by having it nearby the interface that uses it.

The serialization of a model instance used by an interface is not a model concern, it's the interface concern. You'd always want a different serialization mechanism for each interface.

The method is interface-specific (only applicable to one view) - only used in the Wagtail rich text editor. I can't see why we'd want to obscure this by putting it on the model class - we're mixing concerns slightly here.
Serializers generally should be interface-specific and never return anything that's not needed by that interface, rather than thinking of them to be model-specific. It's perfectly normal to have multiple serializers per model, even if it creates duplication it means that you are separating your specific interfaces' code which makes maintenance easier.

```python
# Have this function very nearby so it's clear it's only used in one place.

class PromptDict(TypedDict):
    uuid: uuid.UUID
    label: str
    description: description
    prompt: str
    method: Method

def _serialize_prompt(prompt: Prompt) -> PromptDict:
    # TODO: Make sure you only return fields used by the this specific user-interface.
    return {
            "uuid": self.uuid,
            "label": self.label,
            "description": self.description,
            "prompt": self.prompt_value,
            "method": self.method,
    }

def get_prompts():
    return [_serialize_prompt(prompt) for prompt in Prompt.objects.all()]  
```

We don't want to use an external dependency here at all.

Also, why do we need to change method and uuid to string? Did you try using the Django's encoder? I think it may come with automatic conversion for those.

```python
from django.core.serializers.json import DjangoJSONEncoder

prompts_json = json.dumps(prompts, cls=DjangoJSONEncoder)
```

Using DRF or something else is not necessary and better avoided, especially in a library like this. Very heavy dependencies that are not needed.

Reply:
Thanks for this detailed response Tomasz! it was really insightful